# Data Cleaning

This notebook focuses on identifying and resolving data quality issues discovered during the data understanding phase.

The objective is to prepare a reliable dataset for exploratory data analysis and business intelligence reporting.

In [2]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

pd.set_option('display.max_columns', None)

print("Libraries Loaded Successfully")

Libraries Loaded Successfully


In [3]:
df_2009 = pd.read_excel(
    "../data/raw/online_retail_II.xlsx",
    sheet_name="Year 2009-2010"
)

df_2010 = pd.read_excel(
    "../data/raw/online_retail_II.xlsx",
    sheet_name="Year 2010-2011"
)

df = pd.concat([df_2009, df_2010], ignore_index=True)

In [4]:
print(f"Dataset Shape: {df.shape}")

Dataset Shape: (1067371, 8)


In [5]:
df.head()

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
0,489434,85048,15CM CHRISTMAS GLASS BALL 20 LIGHTS,12,2009-12-01 07:45:00,6.95,13085.0,United Kingdom
1,489434,79323P,PINK CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
2,489434,79323W,WHITE CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
3,489434,22041,"RECORD FRAME 7"" SINGLE SIZE",48,2009-12-01 07:45:00,2.10,13085.0,United Kingdom
4,489434,21232,STRAWBERRY CERAMIC TRINKET BOX,24,2009-12-01 07:45:00,1.25,13085.0,United Kingdom


In [6]:
df.duplicated().sum()

np.int64(34335)

## Duplicate Records Assessment

Duplicate records were identified in the dataset.

Before removing them, a detailed investigation is required to determine whether they represent true duplicates or legitimate repeated transactions.

In [7]:
duplicate_rows = df[df.duplicated()]

duplicate_rows.head(10)

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
371,489517,21912,VINTAGE SNAKES & LADDERS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
383,489517,22130,PARTY CONE CHRISTMAS DECORATION,6,2009-12-01 11:34:00,0.85,16329.0,United Kingdom
384,489517,22319,HAIRCLIPS FORTIES FABRIC ASSORTED,12,2009-12-01 11:34:00,0.65,16329.0,United Kingdom
385,489517,21913,VINTAGE SEASIDE JIGSAW PUZZLES,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
386,489517,21821,GLITTER STAR GARLAND WITH BELLS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
390,489517,84951A,S/4 PISTACHIO LOVEBIRD COASTERS,1,2009-12-01 11:34:00,2.55,16329.0,United Kingdom
391,489517,21491,SET OF THREE VINTAGE GIFT WRAPS,1,2009-12-01 11:34:00,1.95,16329.0,United Kingdom
394,489517,21912,VINTAGE SNAKES & LADDERS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
657,489529,22028,PENNY FARTHING BIRTHDAY CARD,12,2009-12-01 11:51:00,0.42,17984.0,United Kingdom
658,489529,22036,DINOSAUR BIRTHDAY CARD,12,2009-12-01 11:51:00,0.42,17984.0,United Kingdom


In [8]:
duplicate_rows.shape

(34335, 8)

In [9]:
df[df.duplicated(keep=False)].sort_values(
    by=["Invoice", "StockCode"]
).head(20)

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
379,489517,21491,SET OF THREE VINTAGE GIFT WRAPS,1,2009-12-01 11:34:00,1.95,16329.0,United Kingdom
391,489517,21491,SET OF THREE VINTAGE GIFT WRAPS,1,2009-12-01 11:34:00,1.95,16329.0,United Kingdom
365,489517,21821,GLITTER STAR GARLAND WITH BELLS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
386,489517,21821,GLITTER STAR GARLAND WITH BELLS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
363,489517,21912,VINTAGE SNAKES & LADDERS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
371,489517,21912,VINTAGE SNAKES & LADDERS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
394,489517,21912,VINTAGE SNAKES & LADDERS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
362,489517,21913,VINTAGE SEASIDE JIGSAW PUZZLES,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
385,489517,21913,VINTAGE SEASIDE JIGSAW PUZZLES,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
368,489517,22130,PARTY CONE CHRISTMAS DECORATION,6,2009-12-01 11:34:00,0.85,16329.0,United Kingdom


## Duplicate Records Investigation

A detailed review of duplicate records indicates that many rows are exact copies across all available fields.

These records likely resulted from data entry, system processing, or data integration issues rather than legitimate business transactions.

Therefore, duplicate records will be removed during the data cleaning process.

In [10]:
df_clean=df.drop_duplicates()

print(f"Orignal shape: {df.shape}")
print(f"Cleaned shape: {df_clean.shape}")
print("removed rows: ", df.shape[0] - df_clean.shape[0])

Orignal shape: (1067371, 8)
Cleaned shape: (1033036, 8)
removed rows:  34335


In [11]:
df_clean.duplicated().sum()

np.int64(0)

## Issue:
34,335 duplicate records detected.

Investigation:
Duplicate rows were found to be exact copies across all columns.

Decision:
Remove duplicate records.

Implementation:
df_clean = df.drop_duplicates()

In [12]:
df_clean.shape

(1033036, 8)

In [13]:
df_clean.isnull().sum()

Invoice             0
StockCode           0
Description      4275
Quantity            0
InvoiceDate         0
Price               0
Customer ID    235151
Country             0
dtype: int64

In [14]:
df_clean[df_clean["Description"].isnull()].head(20)

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
470,489521,21646,NaN,-50,2009-12-01 11:44:00,0.0,NaN,United Kingdom
3114,489655,20683,NaN,-44,2009-12-01 17:26:00,0.0,NaN,United Kingdom
3161,489659,21350,NaN,230,2009-12-01 17:39:00,0.0,NaN,United Kingdom
3731,489781,84292,NaN,17,2009-12-02 11:45:00,0.0,NaN,United Kingdom
4296,489806,18010,NaN,-770,2009-12-02 12:42:00,0.0,NaN,United Kingdom
4566,489821,85049G,NaN,-240,2009-12-02 13:25:00,0.0,NaN,United Kingdom
6378,489882,35751C,NaN,12,2009-12-02 16:22:00,0.0,NaN,United Kingdom
6555,489898,79323G,NaN,954,2009-12-03 09:40:00,0.0,NaN,United Kingdom
6576,489901,21098,NaN,-200,2009-12-03 09:47:00,0.0,NaN,United Kingdom
6581,489903,21166,NaN,48,2009-12-03 09:57:00,0.0,NaN,United Kingdom


In [15]:
missing_desc = df_clean[df_clean["Description"].isnull()]

missing_desc[["Quantity", "Price"]].describe()

,Quantity,Price
count,4275.000000,4275.0
mean,-17.366784,0.0
std,515.938734,0.0
min,-9600.000000,0.0
25%,-28.000000,0.0
50%,-4.000000,0.0
75%,3.000000,0.0
max,9600.000000,0.0


In [16]:
missing_desc.shape

(4275, 8)

### Missing Description Investigation

Rows with missing product descriptions were examined.

Findings:

- Description is missing.
- Customer ID is also missing.
- Price is zero.
- Many records contain unusual inventory adjustments.

These records do not represent meaningful customer transactions and are unlikely to contribute to sales or customer behavior analysis.

Decision:
Remove rows with missing Description values.

In [17]:
df_clean = df_clean.dropna(subset=["Description"])

In [18]:
df_clean["Description"].isnull().sum()

np.int64(0)

In [19]:
df_clean.shape

(1028761, 8)

In [20]:
df_clean[df_clean["Price"] <= 0].shape

(1744, 8)

In [21]:
price_zero = df_clean[df_clean["Price"] <= 0]

price_zero.head(20)

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
263,489464,21733,85123a mixed,-96,2009-12-01 10:52:00,0.0,NaN,United Kingdom
283,489463,71477,short,-240,2009-12-01 10:52:00,0.0,NaN,United Kingdom
284,489467,85123A,21733 mixed,-192,2009-12-01 10:53:00,0.0,NaN,United Kingdom
3162,489660,35956,lost,-1043,2009-12-01 17:43:00,0.0,NaN,United Kingdom
3168,489663,35605A,damages,-117,2009-12-01 18:02:00,0.0,NaN,United Kingdom
4538,489820,21133,invcd as 84879?,-720,2009-12-02 13:23:00,0.0,NaN,United Kingdom
4674,489825,22076,6 RIBBONS EMPIRE,12,2009-12-02 13:34:00,0.0,16126.0,United Kingdom
5904,489861,DOT,DOTCOM POSTAGE,1,2009-12-02 14:50:00,0.0,NaN,United Kingdom
6556,489899,79323GR,sold as gold,-954,2009-12-03 09:41:00,0.0,NaN,United Kingdom
6781,489998,48185,DOOR MAT FAIRY CAKE,2,2009-12-03 11:19:00,0.0,15658.0,United Kingdom


In [22]:
price_zero[["Quantity", "Price"]].describe()

,Quantity,Price
count,1744.000000,1744.000000
mean,-140.084289,-90.984025
std,870.936457,1939.515507
min,-9600.000000,-53594.360000
25%,-61.250000,0.000000
50%,1.000000,0.000000
75%,2.000000,0.000000
max,12540.000000,0.000000


In [23]:
price_zero["Invoice"].astype(str).str.startswith("C").value_counts()

Invoice
False    1744
Name: count, dtype: int64

In [24]:
price_zero.head(20)

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
263,489464,21733,85123a mixed,-96,2009-12-01 10:52:00,0.0,NaN,United Kingdom
283,489463,71477,short,-240,2009-12-01 10:52:00,0.0,NaN,United Kingdom
284,489467,85123A,21733 mixed,-192,2009-12-01 10:53:00,0.0,NaN,United Kingdom
3162,489660,35956,lost,-1043,2009-12-01 17:43:00,0.0,NaN,United Kingdom
3168,489663,35605A,damages,-117,2009-12-01 18:02:00,0.0,NaN,United Kingdom
4538,489820,21133,invcd as 84879?,-720,2009-12-02 13:23:00,0.0,NaN,United Kingdom
4674,489825,22076,6 RIBBONS EMPIRE,12,2009-12-02 13:34:00,0.0,16126.0,United Kingdom
5904,489861,DOT,DOTCOM POSTAGE,1,2009-12-02 14:50:00,0.0,NaN,United Kingdom
6556,489899,79323GR,sold as gold,-954,2009-12-03 09:41:00,0.0,NaN,United Kingdom
6781,489998,48185,DOOR MAT FAIRY CAKE,2,2009-12-03 11:19:00,0.0,15658.0,United Kingdom


In [25]:
price_zero[["Quantity", "Price"]].describe()

,Quantity,Price
count,1744.000000,1744.000000
mean,-140.084289,-90.984025
std,870.936457,1939.515507
min,-9600.000000,-53594.360000
25%,-61.250000,0.000000
50%,1.000000,0.000000
75%,2.000000,0.000000
max,12540.000000,0.000000


## Zero and Negative Price Investigation

Records with zero or negative prices were investigated.

Findings:

- Many records represent damaged, lost, wet, missing, or manually adjusted inventory.
- These records do not represent actual customer purchases.
- They can distort revenue and product performance analysis.

Decision:
Remove records where Price <= 0.

In [26]:
df_clean = df_clean[df_clean["Price"] > 0]

In [27]:
print(df_clean.shape)
print((df_clean["Price"] <= 0).sum())

(1027017, 8)
0


In [28]:
negative_qty = df_clean[df_clean["Quantity"] < 0]

negative_qty.shape


(19103, 8)

In [29]:
negative_qty.head(20)

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
178,C489449,22087,PAPER BUNTING WHITE LACE,-12,2009-12-01 10:33:00,2.95,16321.0,Australia
179,C489449,85206A,CREAM FELT EASTER EGG BASKET,-6,2009-12-01 10:33:00,1.65,16321.0,Australia
180,C489449,21895,POTTING SHED SOW 'N' GROW SET,-4,2009-12-01 10:33:00,4.25,16321.0,Australia
181,C489449,21896,POTTING SHED TWINE,-6,2009-12-01 10:33:00,2.10,16321.0,Australia
182,C489449,22083,PAPER CHAIN KIT RETRO SPOT,-12,2009-12-01 10:33:00,2.95,16321.0,Australia
183,C489449,21871,SAVE THE PLANET MUG,-12,2009-12-01 10:33:00,1.25,16321.0,Australia
184,C489449,84946,ANTIQUE SILVER TEA GLASS ETCHED,-12,2009-12-01 10:33:00,1.25,16321.0,Australia
185,C489449,84970S,HANGING HEART ZINC T-LIGHT HOLDER,-24,2009-12-01 10:33:00,0.85,16321.0,Australia
186,C489449,22090,PAPER BUNTING RETRO SPOTS,-12,2009-12-01 10:33:00,2.95,16321.0,Australia
196,C489459,90200A,PURPLE SWEETHEART BRACELET,-3,2009-12-01 10:44:00,4.25,17592.0,United Kingdom


In [30]:
negative_qty["Invoice"].astype(str).str.startswith("C").value_counts()

Invoice
True    19103
Name: count, dtype: int64

In [31]:
df.duplicated().sum()

np.int64(34335)

In [36]:
df["Quantity"].describe()

count    1.042727e+06
mean     1.101911e+01
std      1.276894e+02
min      1.000000e+00
25%      1.000000e+00
50%      3.000000e+00
75%      1.000000e+01
max      8.099500e+04
Name: Quantity, dtype: float64

In [39]:
# Remove duplicates
df = df.drop_duplicates()

# Remove inventory adjustment records
df = df[
    ~(
        df["Description"].isna() &
        df["Customer ID"].isna() &
        (df["Price"] == 0)
    )
]

# Remove cancellations
df = df[
    ~df["Invoice"].astype(str).str.startswith("C")
]

# Remove negative or zero quantities
df = df[df["Quantity"] > 0]

# Remove zero/negative prices
df = df[df["Price"] > 0]

In [40]:
df["Revenue"] = df["Quantity"] * df["Price"]

In [43]:
df.describe()


,Quantity,InvoiceDate,Price,Customer ID,Revenue
count,1.007913e+06,1007913,1.007913e+06,779425.000000,1.007913e+06
mean,1.111718e+01,2011-01-04 10:34:18.759594752,4.074252e+00,15320.360461,2.031550e+01
min,1.000000e+00,2009-12-01 07:45:00,1.000000e-03,12346.000000,1.000000e-03
25%,1.000000e+00,2010-07-06 11:21:00,1.250000e+00,13971.000000,4.130000e+00
50%,4.000000e+00,2010-12-09 15:29:00,2.100000e+00,15247.000000,1.008000e+01
75%,1.200000e+01,2011-07-28 12:29:00,4.130000e+00,16794.000000,1.770000e+01
max,8.099500e+04,2011-12-09 12:50:00,2.511109e+04,18287.000000,1.684696e+05
std,1.284700e+02,NaN,5.043045e+01,1695.692775,2.057158e+02


In [42]:
df.isnull().sum()

Invoice             0
StockCode           0
Description         0
Quantity            0
InvoiceDate         0
Price               0
Customer ID    228488
Country             0
Revenue             0
dtype: int64

In [44]:
df.shape

(1007913, 9)

In [45]:
customer_df = df.dropna(subset=['Customer ID']).copy()

In [46]:
df.nlargest(10, 'Quantity')[
    ['Invoice','StockCode','Description','Quantity','Price','Revenue']
]

,Invoice,StockCode,Description,Quantity,Price,Revenue
1065882,581483,23843,"PAPER CRAFT , LITTLE BIRDIE",80995,2.08,168469.6
587080,541431,23166,MEDIUM CERAMIC TOP STORAGE JAR,74215,1.04,77183.6
90857,497946,37410,BLACK AND WHITE PAISLEY FLOWER MUG,19152,0.10,1915.2
127166,501534,21099,SET/6 STRAWBERRY PAPER CUPS,12960,0.10,1296.0
127168,501534,21091,SET/6 WOODLAND PAPER PLATES,12960,0.10,1296.0
127169,501534,21085,SET/6 WOODLAND PAPER CUPS,12744,0.10,1274.4
127167,501534,21092,SET/6 STRAWBERRY PAPER PLATES,12480,0.10,1248.0
135027,502269,21984,PACK OF 12 PINK PAISLEY TISSUES,10000,0.25,2500.0
135028,502269,21982,PACK OF 12 SUKI TISSUES,10000,0.25,2500.0
135029,502269,21980,PACK OF 12 RED SPOTTY TISSUES,10000,0.25,2500.0


In [47]:
df["Quantity"].min()

df["Price"].min()

df.nlargest(10,"Revenue")

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country,Revenue
1065882,581483,23843,"PAPER CRAFT , LITTLE BIRDIE",80995,2011-12-09 09:15:00,2.08,16446.0,United Kingdom,168469.60
587080,541431,23166,MEDIUM CERAMIC TOP STORAGE JAR,74215,2011-01-18 10:01:00,1.04,12346.0,United Kingdom,77183.60
748132,556444,22502,PICNIC BASKET WICKER 60 PIECES,60,2011-06-10 15:28:00,649.50,15098.0,United Kingdom,38970.00
241827,512771,M,Manual,1,2010-06-17 16:53:00,25111.09,NaN,United Kingdom,25111.09
432176,530715,84347,ROTATING SILVER ANGELS T-LIGHT HLDR,9360,2010-11-04 11:36:00,1.69,15838.0,United Kingdom,15818.40
517955,537632,AMAZONFEE,AMAZON FEE,1,2010-12-07 15:08:00,13541.33,NaN,United Kingdom,13541.33
825443,A563185,B,Adjust bad debt,1,2011-08-12 14:50:00,11062.06,NaN,United Kingdom,11062.06
135013,502263,M,Manual,1,2010-03-23 15:22:00,10953.50,12918.0,United Kingdom,10953.50
135015,502265,M,Manual,1,2010-03-23 15:28:00,10953.50,NaN,United Kingdom,10953.50
342147,522796,M,Manual,1,2010-09-16 15:12:00,10468.80,NaN,United Kingdom,10468.80


In [48]:
special_codes = [
    "M",
    "B",
    "AMAZONFEE",
    "POST",
    "DOT",
    "CRUK"
]

df[df["StockCode"].isin(special_codes)]

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country,Revenue
89,489439,POST,POSTAGE,3,2009-12-01 09:28:00,18.00,12682.0,France,54.00
126,489444,POST,POSTAGE,1,2009-12-01 09:55:00,141.00,12636.0,USA,141.00
173,489447,POST,POSTAGE,1,2009-12-01 10:10:00,130.00,12362.0,Belgium,130.00
625,489526,POST,POSTAGE,6,2009-12-01 11:50:00,18.00,12533.0,Germany,108.00
1244,489557,POST,POSTAGE,4,2009-12-01 12:52:00,18.00,12490.0,France,72.00
...,...,...,...,...,...,...,...,...,...
1067001,581498,DOT,DOTCOM POSTAGE,1,2011-12-09 10:26:00,1714.17,NaN,United Kingdom,1714.17
1067191,581570,POST,POSTAGE,1,2011-12-09 11:59:00,18.00,12662.0,Germany,18.00
1067228,581574,POST,POSTAGE,2,2011-12-09 12:09:00,18.00,12526.0,Germany,36.00
1067229,581578,POST,POSTAGE,3,2011-12-09 12:16:00,18.00,12713.0,Germany,54.00


Extreme values were inspected and retained because they represent possible large legitimate transactions.

In [49]:
customer_df = df.dropna(subset=["Customer ID"])

Customer ID missing values were retained in the master dataset because sales transactions remain valid. A separate customer-level dataset will be created later for customer analytics.

## Cleaning Summary

- Removed duplicate records
- Removed rows with missing descriptions
- Removed cancelled transactions
- Removed transactions with non-positive quantity
- Created Revenue feature
- Verified special transaction codes
- Retained Customer ID nulls for sales analysis
- Retained extreme but potentially valid bulk orders

Final Dataset Shape:
(1007913, 9)

In [50]:
df.shape

(1007913, 9)

In [52]:
df.to_csv("../data/processed/online_retail_cleaned.csv", index=False)